---
layout: slide-deck
title:  "STFC Talk"
author: Will Rowe
permalink: /STFC-talk-jan2017/
group: presentation
theme: ember
transition: slide

---


<script type="text/template">

STFC Presentation - 2017

***

# Finding antibiotic resistance genes in metagenomes

***

Will Rowe PhD


<br/>web: [will-rowe.github.io](https://will-rowe.github.io) | email: [will.rowe@liv.ac.uk](mailto:will.rowe@liverpool.ac.uk)

----

## Today's talk

***

* Mining metagenomes for Antibiotic Resistance Genes (ARGs)

* Addressing biological questions

* Current software development

----

## Mining metagenomes for ARGs

***

* Antibiotic resistance is a major public health issue

* Metagenomics is well suited to ARG detection

* Several practical considerations

----

In designing a metagenomic experiment to identify ARGs, what analysis methods and resources are available?

----

![arg-tool-timeline]({{site.url}}/slides/slide-data/host-microbe-talk-jan2017/arg-tool-timeline.svg)

----

### Search Engine for Antimicrobial Resistance (SEAR)

***

* clusters an ARG database to collapse closely related sequences

* clusters reads to representative reference sequences

* maps reads within each cluster & generates consensus sequence

* calculates relative abundance using Reads Per Kilobase / Million (RPKM)

* blasts consensus sequence to ARG databases

---

{% capture markdown-pt1 %}
### SEAR

***

{% endcapture %}
{% capture markdown-pt2 %}

* SEAR annotates full-length ARGs from metagenomic data

* Command line, web and [BaseSpace](https://basespace.illumina.com/apps/2083081/SEAR-Antibiotic-Resistance) versions

>ROWE, W. et al. 2015. SEAR: A cloud compatible pipeline and Web Interface for Rapidly Detecting Antimicrobial Resistance Genes Directly from Sequence Data PLoS One
{% endcapture %}

{{ markdown-pt1 | markdownify }}
<iframe data-src="https://www.illumina.com/informatics/research/sequencing-data-analysis-management/basespace/basespace-apps/sear-antibiotic-resistance-2083081.html" width="700" height="350" frameborder="0" marginwidth="0" marginheight="0" scrolling="yes" style="border:2px solid #000; margin-bottom:5px; max-width: 100%;" allowfullscreen> </iframe>
{{ markdown-pt2 | markdownify }}

---

![basespace1]({{site.url}}/slides/slide-data/host-microbe-talk-jan2017/basespace2.png)

---

![basespace1]({{site.url}}/slides/slide-data/host-microbe-talk-jan2017/basespace3.png)

---
![basespace1]({{site.url}}/slides/slide-data/host-microbe-talk-jan2017/basespace4.png)

----

How can a tool like SEAR by used to address biological questions?

----

<section data-background="http://exoplanets.phy.cam.ac.uk/Meetings/UKCambridgeRiverCamBanner.jpg"></section>

<section data-background="{{site.url}}/slides/slide-data/stfc/cam-overlay.jpg"></section>

----

What roles do the environment & human activities have in disseminating ARGs?

----

<section data-background="/slides/slide-data/host-microbe-talk-jan2017/exap.svg"></section>

----

### Pilot study: ARGs detected in human & animal effluents using metagenomics

***

![pilot-study-fig-1](https://www.researchgate.net/profile/Will_Rowe/publication/287420147/figure/fig3/AS:453676109176832@1485176221310/Fig-1.png)

> ROWE, W. et al. 2016. Comparative metagenomics reveals a diverse range of antimicrobial resistance genes in effluents entering a river catchment. WaterSciTech

----

### Over-expression of *blaOXA* and *blaGES* in hospital effluents

***

![jac-study-fig-2]({{site.url}}/slides/slide-data/host-microbe-talk-jan2017/jac-2.png)

> ROWE, W. et al. 2017. Overexpression of antibiotic resistance genes in hospital effluents over time. J Antimicrob Chemother dkx017. doi: [10.1093/jac/dkx017](https://doi.org/10.1093/jac/dkx017)

---

![jac-study-fig-3]({{site.url}}/slides/slide-data/host-microbe-talk-jan2017/jac-3.png)

---

![jac-study-fig-1]({{site.url}}/slides/slide-data/host-microbe-talk-jan2017/jac-1.png)

----

## Current software development

***

* Although a useful tool, SEAR has limitations

 - decreased resolution by collapsing reference sequences to representatives

 - redundant processing steps

* SEAR was a first attempt at solving a complex problem

 - my first experience in programming

 - pipeline as opposed to software

* Currently developing an ARG-specific aligner for metagenomics

----

![software-overview]({{site.url}}/slides/slide-data/stfc/figure.svg)

---

![software-overview-p1]({{site.url}}/slides/slide-data/stfc/figure-p1.svg)

---

![software-overview-p2]({{site.url}}/slides/slide-data/stfc/figure-p2.svg)

---

![software-overview-p3]({{site.url}}/slides/slide-data/stfc/figure-p3.svg)

----

### ARG-specific read alignment

***

* Written in Go

 - compiled, statically-typed language

 - makes use of Go's concurrency patterns

* Fast mapping of reads

 - concurrent streaming of fastq reads

 - locality sensitive hashing + database sharding

* Quantifies and sub-types annotated ARGs

 - standard SAM output

 - novel variation graph approach to resolve sequence sub-types

* Application to current RNA-seq projects

----

### Related projects

***

* Contamination detection in RNA-seq samples

* Pseudogene prediction

 - required for a large-scale genomics project

 - also trialling machine learning approach

* Web interfaces


> <strong>Acknowledgements</strong>

> Thanks to authors and my current supervisor. This project consists of work from my PhD and PostDoc, supported by the following funding:

> BBSRC Industrial CASE studentship, GSK grants (x2), University of Liverpool 'Dragon's Den' award

----

## Acknowledgements

***

| University of Cambridge | CEFAS | GSK |
| :---------------------: | :---: | :--: |
| Gareth Pearce | David Verner-Jeffreys | Jim Ryan |
| Duncan Maskell | Craig Baker-Austin | |
| | | |


| Sanger Institute | CU Hospital | University of Liverpool |
| :--------------: | :--------: | :--------------: |
| Trevor Lawley | Christianne Micallef | Jay Hinton |
| Mark Stares | |
| Kate Baker | |

</script>
