---
layout: publication
title:  "AL-BALOOL, H. H., WEBER, D., LIU, Y., WADE, M., GULERIA, K., NAM, P. L., CLAYTON, J., <strong>ROWE, W.</strong>, COXHEAD, J., IRVING, J., ELLIOTT, D. J., HALL, A. G., SANTIBANEZ-KOREF, M. & JACKSON, M. S. 2011. Post-transcriptional exon shuffling events in humans can be evolutionarily conserved and abundant. Genome Research, 21, 1788-99"
author: Will Rowe
group: publication
permalink: /genomeresearch-2011/
---

# Post-transcriptional exon shuffling events in humans can be evolutionarily conserved and abundant

Haya H. Al-Balool, David Weber, Yilei Liu, Mark Wade, Kamlesh Guleria, Pitsien Lang Ping Nam, Jake Clayton, William Rowe, Jonathan Coxhead, Julie Irving, David J. Elliott, Andrew G. Hall, Mauro Santibanez-Koref and Michael S. Jackson

[Genome Research](http://genome.cshlp.org/content/21/11/1788.short)

## Abstract

In silico analyses have established that transcripts from some genes can be processed into RNAs with rearranged exon order relative to genomic structure (post-transcriptional exon shuffling, or PTES). Although known to contribute to transcriptome diversity in some species, to date the structure, distribution, abundance, and functional significance of human PTES transcripts remains largely unknown. Here, using high-throughput transcriptome sequencing, we identify 205 putative human PTES products from 176 genes. We validate 72 out of 112 products analyzed using RT-PCR, and identify additional PTES products structurally related to 61% of validated targets. Sequencing of these additional products reveals GT-AG dinucleotides at >95% of the splice junctions, confirming that they are processed by the spliceosome. We show that most PTES transcripts are expressed in a wide variety of human tissues, that they can be polyadenylated, and that some are conserved in mouse. We also show that they can extend into 5′ and 3′ UTRs, consistent with formation via trans-splicing of independent pre-mRNA molecules. Finally, we use real-time PCR to compare the abundance of PTES exon junctions relative to canonical exon junctions within the transcripts from seven genes. PTES exon junctions are present at <0.01% to >90% of the levels of canonical junctions, with transcripts from MAN1A2, PHC3, TLE4, and CDK13 exhibiting the highest levels. This is the first systematic experimental analysis of PTES in human, and it suggests both that the phenomenon is much more widespread than previously thought and that some PTES transcripts could be functional.
