---
layout: publication
title:  "<strong>ROWE, W. P. M.</strong>, VERNER-JEFFREYS, D., BAKER-AUSTIN, C., RYAN, J. J., MASKELL, D. & PEARCE, G. 2016. Comparative metagenomics reveals a diverse range of antimicrobial resistance genes in effluents entering a river catchment. Water Science and Technology, 73, 7, 1541-9"
author: Will Rowe
group: publication
permalink: /wst-2016/
---

# Comparative metagenomics reveals a diverse range of antimicrobial resistance genes in effluents entering a river catchment

Will P. M. Rowe, David Verner-Jeffreys, Craig Baker-Austin, Jim J. Ryan, Duncan Maskell, Gareth Pearce

[Water Science and Technology](http://wst.iwaponline.com/content/73/7/1541)

## Abstract

The aquatic environment has been implicated as a reservoir for antimicrobial resistance genes (ARGs). In order to identify sources that are contributing to these gene reservoirs, it is crucial to assess effluents that are entering the aquatic environment. Here we describe a metagenomic assessment for two types of effluent entering a river catchment. We investigated the diversity and abundance of resistance genes, mobile genetic elements (MGEs) and pathogenic bacteria. Findings were normalised to a background sample of river source water. Our results show that effluent contributed an array of genes to the river catchment, the most abundant being tetracycline resistance genes tetC and tetW from farm effluents and the sulfonamide resistance gene sul2 from wastewater treatment plant (WWTP) effluents. In nine separate samples taken across three years we found 53 different genes conferring resistance to 7 classes of antimicrobial. Compared to the background sample taken up river from effluent entry, the average abundance of genes was three times greater in the farm effluent and two times greater in the WWTP effluent. We conclude that effluents disperse ARGs, MGEs and pathogenic bacteria within a river catchment, thereby contributing to environmental reservoirs of ARGs.
